source/beamline/BioXAS/BioXASSideTableCoordinator.cpp

```diff
@@ -0,0 +1,414 @@
+		verticalUpstreamControl_->move(targetUpstreamHeight);
+	}
+	if (!verticalDownstreamControl_->withinTolerance(targetDownstreamHeight)) {
+		verticalDownstreamControl_->move(targetDownstreamHeight);
+	}
+}
+
+void BioXASSideTableCoordinator::onHeightStopControlValueChanged(double value)
+{
+	Q_UNUSED(value)
+	if(!connectedOnce_)
+		return;
+
+	if (softIOCHeightStopControl_->withinTolerance(1)) {
+		verticalUpstreamControl_->stop();
+		verticalDownstreamControl_->stop();
+
+		softIOCHeightStopControl_->move(0); // restore the value to 0
+	}
+}
+
+void BioXASSideTableCoordinator::onPitchControlValueChanged(double angle)
+{
+	Q_UNUSED(angle)
+	if(!connectedOnce_)
+		return;
+
+	double delta = tableVerticalMotorPosition_ * tan(degreeToRadian(angle));
+
+	// downstream control will move in the direction of the angle and upstream will move against the direction of the angle
+	double targetUpstreamHeight = softIOCHeightControl_->value() - delta ;
+	double targetDownstreamHeight = softIOCHeightControl_->value() + delta;
+
+	if (!verticalUpstreamControl_->withinTolerance(targetUpstreamHeight)) {
+		verticalUpstreamControl_->move(targetUpstreamHeight);
+	}
+	if (!verticalDownstreamControl_->withinTolerance(targetDownstreamHeight)) {
+		verticalDownstreamControl_->move(targetDownstreamHeight);
+	}
+}
+
+void BioXASSideTableCoordinator::onPitchStopControlValueChanged(double value)
+{
+	Q_UNUSED(value)
+	if(!connectedOnce_)
+		return;
+
+	if (softIOCPitchStopControl_->withinTolerance(1)) {
+		verticalUpstreamControl_->stop();
+		verticalDownstreamControl_->stop();
+
+		softIOCPitchStopControl_->move(0); // restore the value to 0
+	}
+}
+
+void BioXASSideTableCoordinator::onLateralControlValueChanged(double value)
+{
+	Q_UNUSED(value)
+	if(!connectedOnce_)
+		return;
+
+	double delta = value - softIOCLateralFeedbackControl_->value();
+
+	double horizontalUpstreamPosition = horizontalUpstreamControl_->value() - delta;
+	double horizontalDownstreamPosition = horizontalDownstreamControl_->value() + delta;
+
+	if (!horizontalUpstreamControl_->withinTolerance(horizontalUpstreamPosition)) {
+		horizontalUpstreamControl_->move(horizontalUpstreamPosition);
+	}
+	if (!horizontalDownstreamControl_->withinTolerance(horizontalDownstreamPosition)) {
+		horizontalDownstreamControl_->move(horizontalDownstreamPosition);
+	}
+}
+
+void BioXASSideTableCoordinator::onLateralStopControlValueChanged(double value)
+{
+	Q_UNUSED(value)
+	if(!connectedOnce_)
+		return;
+
+	if (softIOCLateralStopControl_->withinTolerance(1)) {
+		horizontalUpstreamControl_->stop();
+		horizontalDownstreamControl_->stop();
+
+		softIOCLateralStopControl_->move(0); // restore the value to 0
+	}
+}
+
+void BioXASSideTableCoordinator::onYawControlValueChanged(double angle)
+{
+	Q_UNUSED(angle)
+	if(!connectedOnce_)
+		return;
+
+	double delta = tableHorizontalMotorPosition_ * tan(degreeToRadian(angle));
+
+	if (!horizontalUpstreamControl_->withinTolerance(-delta)) {
```

Comment:
This looks like a relative move.  This ```withinTolerance()``` call won't be checking what you think it does @XuetaoLiu-CLS.  Unless there is something else I'm missing.